api/ondemand_vhds_rds_protocol.rst

```diff
@@ -0,0 +1,111 @@
+
+In order to fix this issue, we are implementing the on-demand Virtual Host
+Discovery Service (VHDS). On-demand VHDS uses the delta xDS protocol, which is
+a separate protocol from xDS. Instead of sending all virtual hosts with a route
+config, using VHDS will allow an Envoy instance to subscribe and unsubscribe
+from a list of virtual hosts stored internally in the xDS management server.
+The xDS management server will monitor this list and use it to filter the
+configuration sent to an individual Envoy instance to only contain the
+subscribed virtual hosts.
+
+Joining/Reconnecting
+====================
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the current
+namespace. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Requesting Additional resources
+===============================
+In order to request additional resources, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+to a xDS management server. The xDS management server will then respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+which is followed up by an Ack/Nack
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+from Envoy.
+
+.. figure:: diagrams/delta_rds_request_additional_resources.svg
+   :alt: Delta RDS request additional resources
+
+Virtual Host Discovery Service
+==============================
+In order to support on-demand, an additional protocol, VHDS, will be added.
+This protocol allows a separation of concerns with RDS where RDS is in charge
+of maintaining route configs while VHDS is in charge of communicating virtual
+hosts. In VHDS, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with :ref:`type_url` set to 
+`type.googleapis.com/envoy.api.v2.route.VirtualHost` 
+and :ref:`resource_names <envoy_api_msg_DeltaDiscoveryRequest.resource_names>` 
+set to a list of routeconfig names + the request HTTP "host"
+headers for which it would like configuration. The management server will
+respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+with the :ref:`resources <envoy_api_msg_DeltaDiscoveryResponse.resources>`
+field populated with the specified virtual hosts, the 
+:ref:`name <envoy_api_msg_DeltaDiscoveryResponse.name>`
+field populated with the virtual host name, and the :ref:`alias` field populated with
+the explicit (no wildcard) domains of the virtual host. Future updates to these
+virtual hosts will be sent via spontaneous updates.
+
+Subscribing to Virtual Hosts
+============================
+Envoy will send a
```

Comment:
Can you explain more about how this works w.r.t. the request path? I.e. how requests are held and resumed, what happens if there is no virtual host found?

Reply:
Done in Dmitri's commit.